lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,728 @@
+  memcpy(ProtoQuals, Protocols.data(),
+         sizeof(clang::ObjCProtocolDecl*)*Protocols.size());
+  auto clangType = clangCtx.getObjCObjectType(superclassTy,
+                                              ProtoQuals,
+                                              Protocols.size());
+  return clangCtx.getObjCObjectPointerType(clangType);
+}
+
+clang::QualType
+ClangTypeConverter::visitBuiltinRawPointerType(BuiltinRawPointerType *type) {
+  return ClangASTContext.VoidPtrTy;
+}
+
+clang::QualType
+ClangTypeConverter::visitBuiltinIntegerType(BuiltinIntegerType *type) {
+  auto &clangCtx = ClangASTContext;
+  if (type->getWidth().isPointerWidth()) {
+    return clangCtx.getUIntPtrType();
+  }
+  assert(type->getWidth().isFixedWidth());
+  auto width = type->getWidth().getFixedWidth();
+  if (width == 1)
+    return clangCtx.BoolTy;
+  return clangCtx.getIntTypeForBitwidth(width, /*signed*/ 0);
+}
+
+clang::QualType
+ClangTypeConverter::visitBuiltinFloatType(BuiltinFloatType *type) {
+  auto &clangCtx = ClangASTContext;
+  auto &clangTargetInfo = clangCtx.getTargetInfo();
+  const llvm::fltSemantics *format = &type->getAPFloatSemantics();
+  if (format == &clangTargetInfo.getHalfFormat())
+    return clangCtx.HalfTy;
+  if (format == &clangTargetInfo.getFloatFormat())
+    return clangCtx.FloatTy;
+  if (format == &clangTargetInfo.getDoubleFormat())
+    return clangCtx.DoubleTy;
+  if (format == &clangTargetInfo.getLongDoubleFormat())
+    return clangCtx.LongDoubleTy;
+  llvm_unreachable("cannot translate floating-point format to C");
+}
+
+clang::QualType ClangTypeConverter::visitArchetypeType(ArchetypeType *type) {
+  // We see these in the case where we invoke an @objc function
+  // through a protocol.
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitDynamicSelfType(DynamicSelfType *type) {
+  // Dynamic Self is equivalent to 'instancetype', which is treated as
+  // 'id' within the Objective-C type system.
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitGenericTypeParamType(GenericTypeParamType *type) {
+  // We see these in the case where we invoke an @objc function
+  // through a protocol argument that is a generic type.
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitSugarType(SugarType *type) {
+  return convert(Type(type->getDesugaredType()));
+}
+
+clang::QualType
+ClangTypeConverter::visitType(TypeBase *type) {
+  // We only convert specific types.
+  return clang::QualType();
+}
+
+clang::QualType ClangTypeConverter::visit(Type type) {
+  return static_cast<super *>(this)->visit(type);
+}
+
+clang::QualType ClangTypeConverter::convert(Type type) {
+  auto it = Cache.find(type);
+  if (it != Cache.end())
+    return it->second;
+
+  // Try to do this without making cache entries for obvious cases.
+  if (auto nominal = type->getAs<NominalType>()) {
```

Comment:
Oh, this check might obviate my comments above about handling the obvious cases.  If so, we should just assert that we aren't in those cases in those functions.

Reply:
Added assertions.